datanode/entities/data_spec.go

```diff
@@ -34,7 +34,7 @@ func (s *ExternalDataSourceSpec) ToProto() *datapb.ExternalDataSourceSpec {
 
 func ExternalDataSourceSpecFromProto(spec *datapb.ExternalDataSourceSpec, txHash TxHash, vegaTime time.Time) (*ExternalDataSourceSpec, error) {
 	if spec.Spec != nil {
-		ds, err := DataSourceSpecFromProto(spec.Spec, txHash, vegaTime)
+		ds, err := DataSourceSpecFromProto(spec.Spec.DeepClone(), txHash, vegaTime)
```

Comment:
Not clear on why this change is necessary? this data is not cached and updated afaik

Reply:
In general, when taking a proto variable and getting the entity equivalent thereof, it's just all round preferable for the two to not have any shared state. Seeing as we're passing this `spec.Spec` to a `FomrProto` function, the clone here is not needed. In other places (most notably liquidity provisions) we were assigning a pointer straight from the proto variable, which we shouldn't be doing.

Reply:
this assumes that the proto message will be updated, this is not the case, it is discarded after the entity is created, this is the same pattern throughout datanode, and it effectively achieves the same as clone, adding a clone is unnecessary, for example:
`func DataSourceSpecFromProto(specProto *datapb.DataSourceSpec) *DataSourceSpec {
	return &DataSourceSpec{
		ID:        specProto.Id,
		CreatedAt: specProto.CreatedAt,
		UpdatedAt: specProto.UpdatedAt,
		Config:    DataSourceSpecConfigurationFromProto(specProto.Config),
		Status:    specProto.Status,
	}
}`

Reply:
As I found out while fixing the unit tests for markets_test.go in sqlstore, because entities and protos both pointed to the same `TradableInstrument`, a number of test cases in there essentially boiled down to `assert.Equal(t, market.TradableInstrument, market.TradableInstrument)`. Even if we don't alter any fields on the proto (or entities) that are shared between multiple variables in the code itself, we were clearly relying on flawed unit tests because of the pointer sharing.

If nothing else, these changes have brought this issue to light, and offer a fix so our unit tests are more solid/reliable.

Reply:
ic, yeah in the case of market specifically  it is not cloning the TradableInstrument and holds a pointer.  However, the entity can still be considered immutable (as is the case with all entities), hence why it is acceptable to share the cached instance of entity (and its corresponding TradableInstrument) in this case without the need to clone it.  Having the tests rely on that impl detail is a bit shoddy as it assumes we don't have cache eviction (which we don't currently), the tests would fail if this were to change and would need to do a value check on tradableinstrument instead.